src/common/Finisher.h

```diff
@@ -66,56 +66,62 @@ class Finisher {
  public:
   /// Add a context to complete, optionally specifying a parameter for the complete function.
   void queue(Context *c, int r = 0) {
-    std::unique_lock ul(finisher_lock);
-    bool was_empty = finisher_queue.empty();
-    finisher_queue.push_back(std::make_pair(c, r));
-    if (was_empty) {
-      finisher_cond.notify_one();
+    {
+      const std::lock_guard l{finisher_lock};
+      const bool should_notify = finisher_queue.empty() && !finisher_running;
+      finisher_queue.push_back(std::make_pair(c, r));
+      if (should_notify) {
+	finisher_cond.notify_one();
+      }
     }
+
     if (logger)
       logger->inc(l_finisher_queue_len);
   }
 
   void queue(std::list<Context*>& ls) {
     {
-      std::unique_lock ul(finisher_lock);
```

Comment:
while you're here, can you please make a template out of these `queue` methods. The implementation is exactly the same no matter the container type.

Reply:
I had that idea as well, but for now, I wanted to make the PR not so intrusive, and concentrate only on the obvious optimizations. Merging duplicate code is a different topic.
(My stgit stack has nearly 200 patches which I will feed into PRs soon; for example, I have made ceph-mds twice as fast. I have many more ideas for optimizations, but first I'll await if you're interested in merging. My Ceph kernel patches were rejected, for example, so I'm cautious.)

Reply:
I prepended a template patch and added a few more small optimizations.